File: 0503_Next Greater Element II/NextGreaterElementII.cs
public class Solution {
    public int[] NextGreaterElements(int[] nums) {
        var ans = new int[nums.Length];
        Array.Fill(ans, -1);
        
        var stack = new Stack<int>();
        
        NextGreaterElements(nums, stack, ans);
        NextGreaterElements(nums, stack, ans);
        
        return ans;
    }
    
    private void NextGreaterElements(int[] nums, Stack<int> stack, int[] ans)
    {
        for(int i=0;i<nums.Length;i++)
        {
            while(stack.Count > 0 && nums[stack.Peek()] < nums[i])
            {
                ans[stack.Peek()] = nums[i];
                stack.Pop();
            }
            
            stack.Push(i);
        }
    }
}

File: 0503_Next Greater Element II/NextGreaterElementII.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        
        Stack<Integer> stack = new Stack<>();
        for(int i=n*2-1;i>=0;i--) {
            while(!stack.isEmpty() && nums[i%n] >= stack.peek()) stack.pop();
            ans[i%n] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(nums[i%n]);
        }
        return ans;
    }
}